dotnet/src/Agents/Core/AgentGroupChat.cs

```diff
@@ -0,0 +1,160 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.SemanticKernel.Agents.Chat;
+using Microsoft.SemanticKernel.ChatCompletion;
+
+namespace Microsoft.SemanticKernel.Agents;
+
+/// <summary>
+/// A an <see cref="AgentChat"/> that supports multi-turn interactions.
+/// </summary>
+public sealed class AgentGroupChat : AgentChat
+{
+    private readonly HashSet<string> _agentIds; // Efficient existence test
+    private readonly List<Agent> _agents; // Maintain order (fwiw)
+
+    /// <summary>
+    /// Indicates if completion criteria has been met.  If set, no further
+    /// agent interactions will occur.  Clear to enable more agent interactions.
+    /// </summary>
+    public bool IsComplete { get; set; }
+
+    /// <summary>
+    /// Settings for defining chat behavior.
+    /// </summary>
+    public ChatExecutionSettings? ExecutionSettings { get; set; }
+
+    /// <summary>
+    /// The agents participating in the nexus.
+    /// </summary>
+    public IReadOnlyList<Agent> Agents => this._agents.AsReadOnly();
+
+    /// <summary>
+    /// Add a <see cref="Agent"/> to the nexus.
+    /// </summary>
+    /// <param name="agent">The <see cref="KernelAgent"/> to add.</param>
+    public void AddAgent(Agent agent)
+    {
+        if (this._agentIds.Add(agent.Id))
+        {
+            this._agents.Add(agent);
+        }
+    }
+
+    /// <summary>
+    /// Process a single interaction between a given <see cref="KernelAgent"/> an a <see cref="AgentGroupChat"/>.
+    /// </summary>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests. The default is <see cref="CancellationToken.None"/>.</param>
+    /// <returns>Asynchronous enumeration of messages.</returns>
+    public async IAsyncEnumerable<ChatMessageContent> InvokeAsync([EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        if (this.IsComplete)
+        {
+            yield break;
+        }
+
+        // Use the the count, if defined and positive, otherwise use default maximum (1).
+        var maximumIterations = Math.Max(this.ExecutionSettings?.MaximumIterations ?? 0, ChatExecutionSettings.DefaultMaximumIterations);
+
+        var selectionStrategy = this.ExecutionSettings?.SelectionStrategy;
+        if (selectionStrategy == null)
+        {
+            yield break;
+        }
+
+        for (int index = 0; index < maximumIterations; index++)
+        {
+            // Identify next agent using strategy
+            var agent = await selectionStrategy.Invoke(this.Agents, this.History, cancellationToken).ConfigureAwait(false);
+            if (agent == null)
+            {
+                yield break;
+            }
```

Comment:
Yeah, let's require selection strategy to return a non-null agent and it can throw if it can't meet that contract.

Termination should only be decided by `TerminationStrategy`, as suggested.